### PR TITLE
set now time at query spec level

### PR DESCRIPTION
**Pull request by jlapacik**
This commit adds a `now` field to a query spec as well as `Now` fields to logical and physical plans. These fields represent the current system time that a query will execute with respect to. This commit is in preparation for adding an interface for getting and setting Flux options - in particular the **now** option.

### CODE REVIEW

**Comment by jlapacik** [comment]
The last commit e939a561fcf081d26fc8ff30c074f270b28f897a introduces an interface for retrieving variable/identifier bindings from a specific block in a Flux program. Also for adding new variable bindings to a specific block of the code after it has already been parsed to an AST and then a semantic program.

This interface allows you to add and retrieve variables from any block of a Flux program, but its intended use case is for getting and setting Flux options. The _options block_ is an implicit block of a Flux program, nested directly inside of the top-level _universe block_. All option variables are bound inside of the _options block_. Therefore you can call the methods `Get` and `Set` directly on this block in order to programmatically get and set options.

`Set` and `SetValues` should be called before interpretation of the program.
`Get` and `GetValues` should be called after interpretation of the program.

As follows:

```
globalScope := interpreter.NewScope()
optionScope := globalScope.Nest()
programScope := optionScope.Nest()

optionScope.Set("task", values.NewIntValue(5))

interpreter.Eval(semanticProgram, programScope, optionScope)

optionSpecifiedInScript := optionScope.Get("some_option")
```

File: query/spec.go [hunk 1]
```diff
@@ -12,6 +13,7 @@ type Spec struct {
 	Edges      []Edge             `json:"edges"`
 	Resources  ResourceManagement `json:"resources"`
 
+	now      time.Time
```
Comment (nathanielc) [inline query/spec.go]: We need to `now` value to be public so it can be set externally and as part of the marshalled JSON.

File: query/spec.go [hunk 2]
```diff
@@ -23,6 +25,19 @@ type Edge struct {
 	Child  OperationID `json:"child"`
 }
 
+// SetNow idempotently sets the current system time for a query
+func (q *Spec) SetNow() {
```
Comment (nathanielc) [inline query/spec.go]: We want to treat the Spec as read only once its been created. So the plan should not call this method, but rather just read the Now field of the Spec. And then if it is Zero change the value in the plan struct.

File: query/spec.go [hunk 3]
```diff
@@ -23,6 +25,19 @@ type Edge struct {
...
+	if q.now.IsZero() {
+		q.now = time.Now().UTC()
+	}
+}
+
+// GetNow retrieves the current system time
+// that a query will execute with respect to.
+func (q *Spec) GetNow() time.Time {
+	return q.now
+}
```
Comment (nathanielc) [inline query/spec.go]: Once Now is public there will be no need for a getter.

File: query/spec.go [hunk 4]
```diff
@@ -73,6 +88,8 @@ func (q *Spec) Parents(id OperationID) []*Operation {
 // prepare populates the internal datastructure needed to quickly navigate the query DAG.
 // As a result the query DAG is validated.
 func (q *Spec) prepare() error {
+	q.SetNow()
```
Comment (nathanielc) [inline query/spec.go]: We do not want to mutate the public parts of the spec in prepare.

File: query/plan/physical.go [hunk 5]
```diff
@@ -14,16 +14,18 @@ import (
...
 	Order      []ProcedureID
-	// Results is a list of datasets that are the result of the plan
-	Results map[string]YieldSpec
 
+	// Results is a list of datasets that are the result of the plan
+	Results   map[string]YieldSpec
 	Resources query.ResourceManagement
+
+	// Now represents the relative current time of the plan.
+	Now time.Time
```
Comment (nathanielc) [inline query/plan/physical.go]: Why did these fields get reordered?

File: query/plan/physical.go [hunk 6]
```diff
@@ -61,13 +63,18 @@ func NewPlanner() Planner {
...
+		return nil, errors.New("Now time is not set")
+	}
+
 	p.plan = &PlanSpec{
-		Now:        now,
 		Procedures: make(map[ProcedureID]*Procedure, len(lp.Procedures)),
 		Order:      make([]ProcedureID, 0, len(lp.Order)),
 		Resources:  lp.Resources,
 		Results:    make(map[string]YieldSpec),
+		Now:        now,
```
Comment (nathanielc) [inline query/plan/physical.go]: Why the change here?

File: query/plan/logical.go [hunk 7]
```diff
@@ -49,6 +51,9 @@ func (p *logicalPlanner) Plan(q *query.Spec) (*LogicalPlanSpec, error) {
 	if err != nil {
 		return nil, err
 	}
+	// GetNow must be called after Walk to ensure
+	// the current system time has been set.
+	p.plan.Now = q.GetNow()
```
Comment (nathanielc) [inline query/plan/logical.go]: This is where the logic to check against Zero should live. 

````
p.plan.Now = q.Now
if p.plan.Now.IsZero() {
 p.plan.Now = time.Now().UTC()
}
```

File: query/control/controller.go [hunk 8]
```diff
@@ -99,7 +99,6 @@ func (c *Controller) createQuery(ctx context.Context, orgID platform.ID) *Query
 		},
 		state:     Created,
 		c:         c,
-		now:       time.Now().UTC(),
```
Comment (nathanielc) [inline query/control/controller.go]: We should keep this behavior. The controller places the incoming queries in a queue. We want to capture the time closest to when the query is submitted not when it finally leaves the queue.

So before the query is planned, check if the spec and set this `now` time if is zero.

In other words, store the time we received the query, then before planning, if the spec does not specify a Now time, use the time we received the query as the now time.

File: query/compile.go [hunk 9]
```diff
@@ -150,6 +164,17 @@ func RegisterBuiltInValue(name string, v values.Value) {
 	builtinScope[name] = v
 }
 
+// RegisterBuiltInOption adds the value to the builtin scope.
+func RegisterBuiltInOption(name string, v values.Value) {
+	if finalized {
+		panic(errors.New("already finalized, cannot register builtin option"))
+	}
+	if _, ok := builtinScope[name]; ok {
```
Comment (nathanielc) [inline query/compile.go]: This should check against builtinOptions

File: query/compile.go [hunk 10]
```diff
@@ -101,6 +107,14 @@ func toSpec(stmtVals []values.Value) *Spec {
 type CreateOperationSpec func(args Arguments, a *Administration) (OperationSpec, error)
 
 var builtinScope = make(map[string]values.Value)
+
+// TODO(Josh): Default option values should be registered similarly to built-in
+// functions. Default options should be registered in their own files
+// (or in a single file) using the RegisterBuiltInOption function which will
+// place the resolved option value in the following map.
+//
+// Note: This is how one should register all options EXCEPT for the special "now" option.
```
Comment (nathanielc) [inline query/compile.go]: Why can't the now options be specified here?

Comment (jlapacik) [reply query/compile.go]: It might be able to actually. If it's being registered, that means the query has already been submitted and is executing. So assuming we could set the option value to `time.Now().UTC()` then we could register the now option just like any other option.

**Comment by jlapacik** [comment]
Variable scoping is a concept relevant only to the interpreter. So how about we encapsulate the notion of scope completely inside the interpreter? We can then add a `GetOption` and `SetOption` interface to the interpreter itself instead of the scope which makes more sense logically and is also a little cleaner.

File: query/interpreter/interpreter.go [hunk 11]
```diff
@@ -10,19 +10,73 @@ import (
...
+	return itrp.globals
+}
+
+// SetVar adds a variable binding to the global scope
+func (itrp *Interpreter) SetVar(name string, val values.Value) {
+	itrp.globals.Set(name, val)
+}
+
+// GetValues returns the evaluated expressions of a Flux program
+func (itrp *Interpreter) GetValues() []values.Value {
```
Comment (nathanielc) [inline query/interpreter/interpreter.go]: Getters in Go should not use the `Get` prefix.

File: query/compile.go [hunk 12]
```diff
@@ -35,7 +35,7 @@ type options struct {
 }
 
 // Compile evaluates a Flux script producing a query Spec.
-func Compile(ctx context.Context, q string, opts ...Option) (*Spec, error) {
+func Compile(ctx context.Context, q string, now time.Time, opts ...Option) (*Spec, error) {
```
Comment (nathanielc) [inline query/compile.go]: Add comment to that `now` must be non zero.

File: query/interpreter/interpreter.go (hunk 11 above)
Comment (nathanielc) [reply query/interpreter/interpreter.go]: Rename `GetValues` to `SideEffects`.

File: query/compile.go [hunk 13]
```diff
@@ -49,27 +49,46 @@ func Compile(ctx context.Context, q string, opts ...Option) (*Spec, error) {
...
 		return nil, err
 	}
 
-	operations, err := interpreter.Eval(semProg, interpScope)
-	if err != nil {
+	if err := itrp.Eval(semProg); err != nil {
 		return nil, err
 	}
-	spec := toSpec(operations)
+	spec := toSpec(itrp.GetValues())
```
Comment (nathanielc) [inline query/compile.go]: `toSpec` should access the intepreter and grab the sideeffects and now option itself.

File: query/compile.go [hunk 14]
```diff
@@ -369,20 +405,18 @@ func DefaultFunctionSignature() semantic.FunctionSignature {
 	}
 }
 
-func BuiltIns() (map[string]values.Value, semantic.DeclarationScope) {
+func BuiltIns() (*interpreter.Interpreter, semantic.DeclarationScope) {
```
Comment (nathanielc) [inline query/compile.go]: Why was this API changed? The uses of BuiltIns all use the scope anyways and no other interperter methods.

File: query/control/controller.go [hunk 15]
```diff
@@ -110,10 +110,14 @@ func (c *Controller) compileQuery(q *Query, queryStr string) error {
...
 	}
-	spec, err := query.Compile(q.compilingCtx, queryStr, query.Verbose(c.verbose))
+	spec, err := query.Compile(q.compilingCtx, queryStr, q.now, query.Verbose(c.verbose))
 	if err != nil {
 		return errors.Wrap(err, "failed to compile query")
 	}
+	// Query has already been created so we are guaranteed to have a now time
+	if spec.Now.IsZero() {
+		spec.Now = q.now
+	}
```
Comment (nathanielc) [inline query/control/controller.go]: This is no longer necessary, we passed in `q.now` so spec.Now is already set.

File: query/functions/join.go [hunk 16]
```diff
@@ -127,7 +127,7 @@ func createJoinOpSpec(args query.Arguments, a *query.Administration) (query.Oper
 				return
 			}
 			p := t.(*query.TableObject)
-			joinParams.add(k/*parameter*/, p/*argument*/)
+			joinParams.add(k /*parameter*/, p /*argument*/)
```
Comment (nathanielc) [inline query/functions/join.go]: Remove comments here.

File: query/interpreter/interpreter.go [hunk 17]
```diff
@@ -10,19 +10,73 @@ import (
...
+
+// GetOption returns a Flux option by name
+func (itrp *Interpreter) GetOption(name string) values.Value {
+	return itrp.options.Get(name)
+}
+
+// GetOptions returns all Flux options
+func (itrp *Interpreter) GetOptions() map[string]values.Value {
+	return itrp.options.GetValues()
+}
```
Comment (nathanielc) [inline query/interpreter/interpreter.go]: Remove unused plural methods.

File: query/compile.go [hunk 18]
```diff
@@ -49,38 +50,70 @@ func Compile(ctx context.Context, q string, opts ...Option) (*Spec, error) {
...
+	}
+
+	for k, v := range builtinOptions {
+		options[k] = v
+	}
+
+	return interpreter.NewInterpreter(options, globals)
+}
+
+func nowFunc(now time.Time) values.Value {
```
Comment (nathanielc) [inline query/compile.go]: This should return `values.Function` i.e the more specific interface.

File: query/compile.go [hunk 19]
```diff
@@ -49,38 +50,70 @@ func Compile(ctx context.Context, q string, opts ...Option) (*Spec, error) {
...
+	name := "now"
+	ftype := semantic.NewFunctionType(semantic.FunctionSignature{
+		ReturnType: semantic.Time,
+	})
+	call := func(args values.Object) (values.Value, error) {
+		return values.NewTimeValue(values.ConvertTime(now)), nil
```
Comment (nathanielc) [inline query/compile.go]: Let's do this conversion only once and then simply return the converted value from this function. 

```
n := values.NewTimeValue(values.ConvertTime(now))
...
return n, nil
```
Or similar

File: query/compile.go [hunk 20]
```diff
@@ -95,12 +128,23 @@ func toSpec(stmtVals []values.Value) *Spec {
 			}
 		}
 	}
+
+	// now option is Time value
+	nowValue, _ := itrp.Option("now").Function().Call(values.NewObject())
```
Comment (nathanielc) [inline query/compile.go]: Can you pass `nil` as the args object?

File: query/compile.go [hunk 21]
```diff
@@ -49,38 +50,70 @@ func Compile(ctx context.Context, q string, opts ...Option) (*Spec, error) {
...
+
+	for k, v := range builtinOptions {
+		options[k] = v
+	}
+
+	return interpreter.NewInterpreter(options, globals)
+}
+
+func nowFunc(now time.Time) values.Value {
+	name := "now"
```
Comment (nathanielc) [inline query/compile.go]: Lets make `now` a constant string so we can refer to it here and below were we read the option. Maybe call it `nowOption`?

File: query/interpreter/interpreter_test.go [hunk 22]
```diff
@@ -2,27 +2,41 @@ package interpreter_test
...
+	_ "github.com/influxdata/platform/query/options"
 	"github.com/influxdata/platform/query/parser"
 	"github.com/influxdata/platform/query/semantic"
 	"github.com/influxdata/platform/query/semantic/semantictest"
 	"github.com/influxdata/platform/query/values"
 )
 
-var testScope = interpreter.NewScope()
+func init() {
+	query.FinalizeBuiltIns()
```
Comment (nathanielc) [inline query/interpreter/interpreter_test.go]: This test should not include any `query` package specific logic.

File: query/interpreter/interpreter_test.go [hunk 23]
```diff
@@ -418,6 +439,67 @@ func TestResolver(t *testing.T) {
 	}
 }
 
+// ExampleInterpreter_Option demonstrates retrieving an option from the Flux interpreter
+func ExampleInterpreter_Option() {
+	itrp := query.NewInterpreter()
```
Comment (nathanielc) [inline query/interpreter/interpreter_test.go]: These examples look good. But they should exist in the `query` package since they are specific to the way the query interpeter is used.

File: query/interpreter/interpreter_test.go [hunk 24]
```diff
@@ -418,6 +439,67 @@ func TestResolver(t *testing.T) {
...
+	// The now option is a function value whose default behavior is to return
+	// the current system time when called.
+	nowTime, _ := nowFunc.Function().Call(values.NewObject())
+	fmt.Fprintf(os.Stderr, "The current system time (UTC) is: %v\n", nowTime)
+	// Output:
+}
+
+// ExampleInterpreter_SetOption demonstrates setting an option from the Flux interpreter
+func ExampleInterpreter_SetOption() {
+	itrp := query.NewInterpreter()
```
Comment (nathanielc) [inline query/interpreter/interpreter_test.go]: This example should probably just use the `inteperter.Interpreter` type, not the `query` specific one.

File: query/interpreter/interpreter_test.go [hunk 25]
```diff
@@ -418,6 +439,67 @@ func TestResolver(t *testing.T) {
...
+	// Set a new option from the interpreter
+	itrp.SetOption("dummy_option", values.NewIntValue(3))
+
+	fmt.Printf("dummy_option = %d", itrp.Option("dummy_option").Int())
+	// Output: dummy_option = 3
+}
+
+// Example_overrideDefaultOption demonstrates a query using an option
+// that is overridden before the query it evaluated by the interpreter.
+func ExampleInterpreter_overrideDefaultOption() {
```
Comment (nathanielc) [inline query/interpreter/interpreter_test.go]: Rename to `ExampleInterpreter_OverrideDefaultOption`

File: query/values/function.go [hunk 26]
```diff
@@ -0,0 +1,89 @@
+package values
+
+import (
+	"regexp"
+
+	"github.com/influxdata/platform/query/semantic"
+)
+
+// NewFunction returns a new function value
+func NewFunction(name string, typ semantic.Type, call func(args Object) (Value, error), sideEffect bool) Function {
```
Comment (nathanielc) [inline query/values/function.go]: Can you move the `Function` interface to this file as well?

File: query/values/function.go [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+func (f function) Array() Array {
+	panic(UnexpectedKind(semantic.Object, semantic.Function))
+}
+
+func (f function) Object() Object {
+	panic(UnexpectedKind(semantic.Object, semantic.Object))
+}
+
+func (f function) Function() Function {
+	return &f
```
Comment (nathanielc) [inline query/values/function.go]: This will create a copy of the function struct each time, which doesn't change the behavior since the struct contains a pointer to a Go function. As a result it is wasteful. Since the implementation requires that a pointer receiver be used on at least one method, a pointer receiver should be used for all methods.

Can you update all method to use a pointer receiver?

File: query/interpreter/interpreter_test.go (hunk 25 above)
Comment (jlapacik) [reply query/interpreter/interpreter_test.go]: Because the Interpreter does not have a method called OverrideDefaultOption, the Go testing example framework says to keep lowercase. The example will actually throw an error otherwise.
